### PR TITLE
WIP: Add non-throwing variants of parse() and to()

### PR DESCRIPTION
**Pull request by nordlow**
Putting it out in the open just to get early feedback.

See also: https://forum.dlang.org/post/hotgouygobclmesixayw@forum.dlang.org

### CODE REVIEW

**Comment by wilzbach** [comment]
I like the idea (and I'm generally in favor of using more `Nullable` / `Optional` in Phobos).
However a few concerns for now:
- should we really promote a different type and not use [`Nullable`](https://dlang.org/phobos/std_typecons.html#Nullable). I'm aware of the problem that Nullable doesn't allow one to store a custom error message. I guess at the very least the API should be similar (maybe we can generalize the Nullable concept like we did with ranges). Also there are lots of nice goodies in `Nullable` like `apply` (the analog of Folly's  `then`)
- `expected.result` (`.get` in Nullable) should automatically `throw` when accessed and I think we might be better off with sth. like `get!ConvException`. Only `Nullable.get(<fallback>)` is `nothrow`

**Comment by nordlow** [comment]
Sounds good to me.

Is `Nullable.apply` nothrow?

**Comment by wilzbach** [comment]
>  Is Nullable.apply nothrow?
> Can we feed Nullable.apply a lambda and still be nothrow?


Yes and even `@nogc` e.g.

https://github.com/dlang/phobos/blob/bb769cfaf51f32a9f20567f885361fe828e9f8a3/std/typecons.d#L3807-L3823

See also: https://dlang.org/changelog/2.080.0.html#std-typecons-nullable-apply

**Comment by nordlow** [comment]
Ok, I'm all in on using `Nullable`. Lets make this happen!

**Comment by jacob-carlborg** [comment]
I don’t think we should use Nullable, instead implement a proper Optional type. I’ve an implementation here [1] that implements the range interface and safe access. 

[1] https://github.com/jacob-carlborg/dlp/blob/master/source/dlp/core/optional.d

**Comment by nordlow** [comment]
Can you briefly describe why we need an Optional instead of a Nullable?

**Comment by jacob-carlborg** [comment]
> Can you please describe why we need an Optional instead of a Nullable? Are there any other reasons than the above mentioned?

No, it's mostly because it has a range interface and safe access interface. I also think it's a better, more descriptive name.

**Comment by nordlow** [comment]
How the is the interface `@safer`? 

AFAICT `Optional.get` is as (un)safe as `Nullable.get` is. Both may result in unexpected behaviour when compiled without asserts.

**Comment by jacob-carlborg** [comment]
> How the is the interface `@safer`?

I didn't say `@safe`, I said "safe access", without the at sign. "safe" might not be the best name but what I mean is you can do this:

```
class Bar { int a; }
class Foo { Bar bar; }
auto foo = none!Foo;
auto a = foo.bar.a; // safe access, "a" will be an optional int
auto b = a.or(3); // unwraps "a" if it has a value, otherwise returns 3
auto c = a.map!(i => i * 4); // it also supports the range interface
```

If the standard algorithms are used, the lambda will only be called if the optional has a value. When using `map` it's basically like `apply` for `Nullable` but with a standard interface.

**Comment by nordlow** [comment]
So lets try adding your `opDispatch` logic to `Nullable` then!

**Comment by ZombineDev** [comment]
How about until we all settle on the return type, we simply use `auto`? That way the exact return type becomes an implementation detail and in future, when we have made the necessary changes to `Nullable` or we have added an new type, we could switch from `auto` to an explicit return type.

**Comment by ZombineDev** [comment]
Another option would be add these functions first to `std.experimental`, so we would be free to change the return type later.

**Comment by jacob-carlborg** [comment]
> So lets try adding your opDispatch to Nullable then!

Please no. I suggest we deprecate `Nullable` and add a new implementation called `Optional`. The name `Nullable` is poorly chosen. The name `Nullable` suggests it should only be used for value types which otherwise cannot be assigned `null`. But `Optional` is supposed to be used for all types, even those that can be assigned `null`. An optional pointer would mean that the pointer might not point to something, a plain pointer would mean that it always points to something, it can never be `null` (I know that this is not enforced by the language).

**Comment by aliak00** [comment]
I agree `Nullable` can be used here but it's says nothing about "this value may be present" in a truely generic sense - null is a very valid part of some value type domains - e.g. pointers. 

An optional is explicit in intent and it'd be a much nicer interface with an Optional. Also, once we start using `Nullable` in phobos for stuff like this it's going to be very hard to go back :/ 

So super big +1 for putting this in std.experimental until we have an Optional type.

**Comment by jacob-carlborg** [comment]
Nothing outside `std.experimental` is allowed to use `std.experimental`. Perhaps create an internal function marked with `package` to be able to share it?

**Comment by nordlow** [comment]
You mean an internal function `_tryParse()` marked with package inside std.experimental.typecons?

**Comment by nordlow** [comment]
Incidentally...: https://forum.dlang.org/post/vyzbihhgrvmiqacakezk@forum.dlang.org

**Comment by wilzbach** [comment]
> Nothing outside std.experimental is allowed to use std.experimental.

Nope, that's not true. You can't return experimental types, but Phobos __is allowed__ to use things from experimental internally. We do so with `checkedint` and the `allocator` packages in a few places already.

**Comment by nordlow** [comment]
So, @wilzbach and @jacob-carlborg...should I put `tryParse`and `tryTo`

- in `std.experimental` and call them from `std.typecons` or
- put it `std.typecons` and mark the new functions with `package`?

**Comment by jacob-carlborg** [comment]
I will let @wilzbach decide.

**Comment by wilzbach** [comment]
> I will let @wilzbach decide.

I would go with `experimental` because they are then at least partially exposed whereas `package` functions are entirely hidden from the user.

**Comment by wilzbach** [comment]
> You mean an internal function. say tryParse(), marked with package inside std.experimental.typecons?

Why `stdx.typecons` and not `std.experimental.conv`?

**Comment by nordlow** [comment]
Ahh, my mistake..,of course they should go into `std.experimental.conv`.

**Comment by jacob-carlborg** [comment]
> I would go with experimental because they are then at least partially exposed whereas package functions are entirely hidden from the user.

What I meant was to put all the logic in a package function inside `std.conv` and expose it in `std.experimental.conv.tryParse` and `std.conv.parse`.

File: std/conv.d [hunk 1]
```diff
@@ -2636,21 +2637,7 @@ do
...
- *
- * Returns:
- *     An `enum` of type `Target`
- *
- * Throws:
- *     A $(LREF ConvException) if type `Target` does not have a member
- *     represented by `s`.
- */
-Target parse(Target, Source)(ref Source s)
+Nullable!(Target) tryParse(Target, Source)(ref Source s)
```
Comment (andralex) [inline std/conv.d]: Initiative is good but I think we'd need an `Expected` facility to make it right: https://www.youtube.com/watch?v=nVzgkepAg5Y

Comment (aliak00) [reply std/conv.d]: With a `noexcept` Expected facility this would be even better!
We can do this with SumType: https://run.dlang.io/is/82Q0DR

**Comment by FeepingCreature** [comment]
I have only one plea.

If we introduce an `Optional` as a possible alternative to `Nullable`. Then whatever you do, _please for the love of God_ do not `alias get this`! It was an error with Nullable, though I suppose it did introduce feature parity with pointers in the field of unexpected crashes...

**Comment by jacob-carlborg** [comment]
Yeah, `get` should be rarely used.

**Comment by jmdavis** [comment]
> Yeah, get should be rarely used.

Actually, I think that his point was the opposite. Rather, `get` should be used explicitly to access the value rather than having it implicitly convert to the type it's wrapping, because the `alias this` makes it very easy to accidentally convert to the wrapped type, which of course blows up in your face if the `Nullable` doesn't contain a value. He's brought this up on several occasions previously and wanted to deprecate the `alias this` on `Nullable`, but a compiler bug prevented it.

**Comment by jacob-carlborg** [comment]
Yes, I agree. I meant `get` should not be used, regardless if it's called directly or through `alias this`. Functions like `getOrElse`, `map` or `filter` should be used instead.

**Comment by nordlow** [comment]
I've added a very basic version of `Expected` [here](https://github.com/nordlow/phobos-next/blob/master/src/expected.d). Would you be so kind to add some comments?

**Comment by FeepingCreature** [comment]
I can't seem to add comments to an ordinary file. Could you open a mini-PR and link it here?

**Comment by FeepingCreature** [comment]
The logic of using ranges is, I think, that a type which may either contain one value or no value, ie. Nullable, is sort of a special case of a type which may contain any number of values, ie. ranges.

**Comment by jacob-carlborg** [comment]
> Is it ok to add Expected to this PR for now in order for you to make comments in it?

It would be better to have a separate PR. Would it make sense to build a new Optional type as well and have it be based on each other?

**Comment by FeepingCreature** [comment]
Maybe `VariantN` should be the backing type in any case. I don't know if it has open issues.

**Comment by jmdavis** [comment]
> Optional is not supposed to act like a pointer, in my opinion. Perhaps Nullable is.

Honestly, the way `Nullable` works is pretty much exactly what I'd expect from an `Optional` type except maybe for the issue of `alias this`. I really don't see any difference between a type emulating a pointer and one that just optionally holds a value.

Now, having a type that either holds a value or hold some sort of error type as proposed here _is_ something different, but in its case, it basically holds a value regardless. It's just that it could be the target type or the error type.

> Maybe VariantN should be the backing type in any case. I don't know if it has open issues.

Using std.variant does tend to cause subtle problems, but `union` has at least some of those problems on its own already. So, if the solution is going to contain a `union`, then maybe it would make more sense to use `VariantN`. Personally, when I've created types like `Expected` in the past, I've just used separate variables for the value and the error, since it's simpler and avoids the problems you typically get with unions. However, it does have the downside of making the struct take up more space. I don't know which approach is ultimately better.

**Comment by FeepingCreature** [comment]
> I've just used separate variables for the value and the error

This runs into the problems that I've recently fixed in Nullable (by using a `union` :smile:) involving structs whose `.init` state is not a valid type instance, where destruction of the sum type leads to a false destructor call on the field, possibly causing crashes. So I definitely recommend using a `union`.

**Comment by aliak00** [comment]
A union seems like a pretty good approach to back an Expected type - [here](https://run.dlang.io/is/oeiwXB)'s a dirty proof of concept based on [SumType](https://github.com/pbackus/sumtype) by @pbackus.

> I really don't see any difference between a type emulating a pointer and one that just optionally holds a value.

1) The semantics are completely different. Is "no value" a valid value? (this is a valid question for a pointer type, but not for optional). Who owns the value (valid for pointer, but not for an optional value type, barring indirections which applies to any value type). 2) safety - a pointer does not force you to check for presence in anyway. An optional, depending on how it's implemented, does. 3) intent - what does a pointer return mean? An optional is clear as day - ok this overlaps with 1.

While you can treat a pointer the same sometime, it becomes conventional and not enforceable.

**Comment by FeepingCreature** [comment]
I played with it a bit, and added the Turducken testcase: [code](https://run.dlang.io/is/r6L73L). Seems to eat it fine.

**Comment by pbackus** [comment]
I'm working on [a fleshed-out version of `Expected`](https://github.com/pbackus/expected). Comments welcome (though it's still pretty rough).

File: std/conv.d (hunk 1 above)
Comment (nordlow) [reply std/conv.d:2640]: I believe we should use a Result type. FYI, @pbackus

**Comment by andralex** [comment]
Just a note - returning either Nulltable or Optional from parsing functions makes it impossible to distinguish the cause of error when there could be more than one (e.g. end of stream vs. overflow). With Expected the result contains arbitrarily rich information in case of failure.

**Comment by nordlow** [comment]
> Just a note - returning either Nulltable or Optional from parsing functions makes it impossible to distinguish the cause of error when there could be more than one (e.g. end of stream vs. overflow). With Expected the result contains arbitrarily rich information in case of failure.

Is Expected different from Result?

**Comment by thewilsonator** [comment]
Result is usually
```
struct Result(T) { T thing; bool error; } 
```

whereas expected is more like
```
struct Expected(T)
{
     union { T thing; struct { void* errorObject; TypeInfo ti;} }
     bool error;
}
```

**Comment by pbackus** [comment]
`Result` is typically a union of an expected value and an error value (e.g. in [Rust][1], [F#][2], [OCaml][3]).

`Expected(T, E)` is just another name for `Result(T, E)`; `Expected(T)` is equivalent to `Result(T, E)` with `E` fixed to some specific type, such as `Exception`.

Personally I would recommend against using the name "expected". The most common names for this type in functional-programming languages are "result" and "either".

[1]: https://doc.rust-lang.org/std/result/
[2]: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/results
[3]: https://ocamlbook.org/error-handling/#the-functional-way--option-and-result-types

**Comment by andralex** [comment]
One issue with `Result!(T, E)` is that it commits to the error type, which in our case means we need to use heap allocation for the exception. I wish we had a few more options for the error case:

* `Exception`
* a lambda
* perhaps a string describing the error

Wrt names I find `Result` is better than `Either` because the latter gives equal importance to the branches, whereas `Result` can be easier thought as "normal path and contingency path".

**Comment by nordlow** [comment]
> One issue with `Result!(T, E)` is that it commits to the error type, which in our case means we need to use heap allocation for the exception. I wish we had a few more options for the error case:

How does `Result!(T, E)` limit `E` to be a heap-allocated exception?

**Comment by andralex** [comment]
@nordlow I just meant it limits the error to one type. I think it would be better (without adding other types to Result) to allow not only allocated exceptions but lambdas that throw exceptions and strings from which exceptions can be constructed.

In brief: `Result!(T, Exception)` allocates the exception eagerly even if not thrown. A better approach would be `Result!T` that makes the exception construction lazy as well.

**Comment by pbackus** [comment]
@andralex If we want to store information from which an exception can be constructed, rather than an actual `Exception`, then there's no reason to use `Exception` as the error type. Instead, we can use the type of that information directly; e.g.,

```d
alias ExceptionArgs = Tuple!(string, "msg", string, "file", size_t, "line");

Result!(T, ExceptionArgs) fun(...)
{
    // ...
    if (ok)
        return typeof(return)(T(...));
    else
        return typeof(return)(ExceptionArgs("no good", __FILE__, __LINE__));
}
```

There's plenty of room for polish here, of course, but the basic approach ought to work.

**Comment by andralex** [comment]
@pbackus understood, though exposing the arguments all the way to the user-facing API is not really nice (what if you change them later etc).

Users should have a simple mechanism to get some uniform `Result!T` type they can they inspect - is this a `T`, and if not, what was the problem? Give me an integral error code, or a string description, or an `Exception` object I can throw around.

**Comment by nordlow** [comment]
> Users should have a simple mechanism to get some uniform `Result!T` type they can they inspect - is this a `T`, and if not, what was the problem? Give me an integral error code, or a string description, or an `Exception` object I can throw around.

Note that SIL's `Result` type by design doesn't take the error type as template parameter exactly because of the same reason.

**Comment by pbackus** [comment]
This sounds like a mechanism-vs-policy issue. Certainly it would be ideal if we could live in a world where all functions reported errors in some uniform way, and it might be a good idea to adopt such a policy for Phobos, but we do not live in such a world, and I do not think that a general-purpose `Result` implementation is the right place to encode such a policy. Rather, `Result` should provide the mechanism, and leave the policy to its users.

**Comment by LightBender** [comment]
This idea would be better pursued as a low-level API design for Phobos 3.

**Pull request closed without merging**